### PR TITLE
Better default styles computation.

### LINKED ISSUE #95: Tables where cell padding has been set to 0 lead to cropped/non-matching image
### Use case: description, code  

[jsfiddle](https://jsfiddle.net/4n52yzcg/)

In production, I have a large table and noticed that the bottom and left edges were cut off when saving to Png. Using fiddle I isolated the problem to setting the padding on table cells to 0.

In the fiddle the bottom edge is cut off and also the image is wider than the dom table so I guess there may be various small differences when padding is 0 in this case.

Seems related to https://github.com/1904labs/dom-to-image-more/issues/77 where the user also fixed the issue by adding padding. Tables naturally have 1px cell padding so this might not have been noticed in the general case.

image

### Expected behavior

Image should be full table.

### Actual behavior (stack traces, console logs etc)

Image doesn't match. Edges are cropped etc.

### Library version

2.13.0

### Browsers

- Chrome 108

### LINKED ISSUE #106: Use full hierarchy when determining the browser default styles
### Use case: description, code  

Per the original reported problem of #95, and the discoveries made while moving towards a solution in #104, we know that the only reasonable way to get the full defaults of a particular element is to generate a matching DOM element hierarchy in the sandbox such that (for example) a `TD` in a `TR` in a `TBODY` in a `TABLE` in a `DIV` in `BODY` would need all the intervening elements created in the sandbox and then we could insert the element we need now to determine its in-situ default styles.

### Expected behavior

Insert a `TD` and see the default `1px` style that would have been driven by the fact that it's nested in a `TABLE` with a `TBODY` with a `TR`

### Actual behavior (stack traces, console logs etc.)

On Chrome, they default style of a bare `TD`  inserted as a direct descendent of `BODY` lies and says that its padding is `0px` instead of what would actually be a `1px` padding due to the `table-cell` display.

### Library version

2.13.1 and 3.0.0

### Browsers

- [X] Chrome 49+  
- [ ] Firefox 45+

### PR DESCRIPTION
**Pull request by IDisposable**
Uses the correct tag hierarchy when computing the default style to account for (implicit) browser stylesheets that use tags within a hierarchy to specify default styles.

We create the elements in the sandbox with all the required parent nodes to match up to the first [block-level element](https://developer.mozilla.org/en-US/docs/Web/HTML/Block-level_elements), compute the default styling and then tear down that hierarchy. This work is still only done once per tag-and-parents combination.

Resolves #106 and thus also Resolves #95

### CODE REVIEW

**Comment by IDisposable** [comment]
Hey @meche-gh, this should do the trick... I also refactored that part because it was getting a bit long-winded.

File: src/dom-to-image-more.js [hunk 1]
```diff
@@ -1010,45 +1012,147 @@
...
+        'UL',
+        // these are ultimate stoppers in case something drastic changes in how the DOM works
+        'BODY',
+        'HEAD',
+        'HTML',
+    ];
+
+    function getDefaultStyle(sourceElement) {
+        const tagHierarchy = computeTagHierarchy(sourceElement);
+        const tagKey = tagHierarchy.join('>'); // it's like CSS
```
Comment (meche-gh) [inline src/dom-to-image-more.js:1059]: This adds a cost for generalised tags like span and i which I suspect is exponential - $O(c^N)$ where $c$ is average hierarchy depth for that tag. As we're constructing against tag hierarchy in the DOM, where most developers religiously follow "permitted (non-normative) context", we should be able to get away with caching by tag name, or reducing the key to a maximum of 2 - `tagName` or `ascentStopper tagName`. Need to protect against #70 regression here

Comment (meche-gh) [reply src/dom-to-image-more.js:1059]: The ideal way to stress this algo in performance testing is to go to a complex Wikipedia article like https://en.wikipedia.org/wiki/JavaScript?useskin=vector-2022

If you have qualms about overly simplifying the caching logic, we can loop in the original author and see what they think RE: perf cost-benefit

**Comment by IDisposable** [comment]
That's a fair concern... I ran some tests against some huge tests I have
and found 5 tag depth is pretty much the max before hitting a stopper since
we're not factoring siblings in (and siblings are a cache hit). I suspect
it's not a huge issue, but we would need at least 4 for the fix to kick in
for `TD` cells.

I wonder if we have a reproduction for the #70 issue handy?

Also wondering if we would be better served for now in pulling in your post
filter stuff and maybe putting the optimizations behind a feature flag?

**Comment by meche-gh** [comment]
For `td` - if we cache as `tagName` or `acsentStopper tagName` - the fix should always kick in when TD is wrapped by TABLE, and we get new cache entries for tag abuse e.g. div>td. Plus tags like SPAN and I will probably only produce the permutations `span, div span, p span`

**Comment by IDisposable** [comment]
That may work... we're not going to save anything walking up the parentage,
and we need to create the tags we found. So this would effectively save
caching a few edge cases (and smaller tagKey strings...but that is a tiny
factor). I will test with a bunch of big pages and see what an instrumented
run shows.

**Comment by meche-gh** [comment]
> Also wondering if we would be better served for now in pulling in your post
filter stuff and maybe putting the optimizations behind a feature flag?

I was thinking the filter could be used for the other dom-to-image fork packages. At the same time, I'd imagine a lot of opt-ins. I'll see if I can refactor my secret WIP branch into a neat minimal PR.

**Comment by meche-gh** [comment]
Just to be clear, we do want to be creating keys against the hierarchy tree, and my concern about the perf can be handled by a change to the logic for creating the cache key.

File: src/dom-to-image-more.js [hunk 2]
```diff
@@ -1010,45 +1012,147 @@
...
+        'P',
+        'PRE',
+        'SECTION',
+        'SVG',
+        'TABLE',
+        'UL',
+        // these are ultimate stoppers in case something drastic changes in how the DOM works
+        'BODY',
+        'HEAD',
+        'HTML',
```
Comment (meche-gh) [inline src/dom-to-image-more.js:1054]: [nit] The `svg` addition is ultimately custom - it's a performance boost!
SVG tags inside an `<svg>` are not going to have their default CSS change because of the parent elements outside of their `<svg>` tag:

```suggestion
        'TABLE',
        'UL',
        // these are ultimate stoppers in case something drastic changes in how the DOM works
        'BODY',
        'HEAD',
        'HTML',
        'SVG',
```

Comment (meche-gh) [reply src/dom-to-image-more.js:1054]: You can also add `<math>` as the MathML sectioning root. It is not part of the HTML spec unfortunately, but web browsers irregularly implement it as a separate concurrent web technology

**Review by zm-cttae — CHANGES_REQUESTED** [review]

File: src/dom-to-image-more.js (hunk 1 above)
Comment (IDisposable) [reply src/dom-to-image-more.js:1059]: Yep, that would do it _IFF_ the assertion is that `TD>TR>TBODY>TABLE` _results_ in the same default styles as `TD>TR>TFOOT>TABLE` and other similar, but not sufficiently dissimilar fragments (both would end up with the `tagKey` of `'TD>TABLE'`. I'm going to see what it looks like. Do you have a performance test-jig for doing arbitrary pages? Might be a nice jsfiddle

Comment (IDisposable) [reply src/dom-to-image-more.js:1059]: I've updated for the option to be `strict` or `relaxed` in case someone breaks. We default to `strict`

File: src/dom-to-image-more.js (hunk 2 above)
Comment (IDisposable) [reply src/dom-to-image-more.js:1054]: `math` added, in lower case... I suspect when browsers actually do adopt the tag it'll morph to uppercase...

Comment (meche-gh) [reply src/dom-to-image-more.js:1054]: Cool. We need to put `svg` (another lowercase exception) after the "these are ultimate stoppers" comment.  
Just to be sure it doesn't get deleted later when the list gets updated

Comment (meche-gh) [reply src/dom-to-image-more.js:1054]: - We need to lowercase `svg` also!
- The `svg` tag won't appear in the MDN list so we should move it under the `// these are ultimate stoppers` comment

File: src/dom-to-image-more.js [hunk 3]
```diff
@@ -1010,45 +1012,147 @@
...
+            tagHierarchy
+        );
+        const defaultStyle = computeStyleForDefaults(sandboxWindow, defaultElement);
+        destroyElementHierarchy(defaultElement);
+
+        tagNameDefaultStyles[tagKey] = defaultStyle;
         return defaultStyle;
+
+        function computeTagHierarchy(sourceNode) {
+            const ELEMENT_NODE = 1;
```
[the remark this answers is not in the document]
Comment (IDisposable) [reply src/dom-to-image-more.js:1094]: Pulled those changes in and now have the ability to key the defaultStyleCache with either the fill tag hierachy (`strict` mode) or just `ascentStopper + tagName` ('relaxed` mode). This should make it possible revert if we broke something.

File: src/dom-to-image-more.js (hunk 2 above)
Comment (IDisposable) [reply src/dom-to-image-more.js]: Oh, doh... I can fix that real quick... completely missed what you were going for...
